### PR TITLE
Added parsers for Siemens WinCC log files

### PR DESCRIPTION
**Pull request by rgayon**
## One line description of pull request

Add parsers for two Siemens WinCC log files

## Checklist:
* [x] Automated checks (GitHub Actions, AppVeyor) pass
* [x] No new [new dependencies](https://plaso.readthedocs.io/en/latest/sources/developer/Adding-a-new-dependency.html) are required or l2tdevtools has been updated
* [x] Reviewer assigned

### CODE REVIEW

File: plaso/parsers/wincc.py [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+from dfdatetime import time_elements as dfdatetime_time_elements
+
+from plaso.containers import events
+from plaso.lib import errors
+from plaso.parsers import interface
+from plaso.parsers import manager
+
+
+class SIMATICS7EventData(events.EventData):
+  """SIMATIC S7 event data."""
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:15]: Add attributes to docstring.

File: plaso/parsers/wincc.py [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+  def __init__(self):
+    """Initializes event data."""
+    super(SIMATICS7EventData, self).__init__(data_type=self.DATA_TYPE)
+
+    self.creation_time = None
+    self.message = None
+
+
+class WinCCSysLogEventData(events.EventData):
+  """WinCC Sys Log event data."""
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:28]: Add attributes to docstring.

File: plaso/parsers/wincc.py [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+
+  def __init__(self):
+    """Initializes event data."""
+    super(SIMATICS7EventData, self).__init__(data_type=self.DATA_TYPE)
+
+    self.creation_time = None
+    self.message = None
+
+
+class WinCCSysLogEventData(events.EventData):
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:27]: Win => Windows ? Or is the name of the product WinCC ?

File: plaso/parsers/wincc.py [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+    Args:
+      parser_mediator (ParserMediator): mediates interactions between parsers
+          and other components, such as storage and dfVFS.
+      line_number (int): number of the line the values were extracted from.
+      values (list[str]): values extracted from the line.
+
+    Raises:
+      WrongParser: when the values cannot be parsed.
+    """
+
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:68]: remove empty line

File: plaso/parsers/wincc.py [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+class WinCCSysLogEventData(events.EventData):
+  """WinCC Sys Log event data."""
+
+  DATA_TYPE = 'wincc:sys_log:entry'
+
+  def __init__(self):
+    """Initializes event data."""
+    super(WinCCSysLogEventData, self).__init__(data_type=self.DATA_TYPE)
+
+    self.log_id = None
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:36]: id => identifier

File: plaso/parsers/wincc.py [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+    self.creation_time = None
+    self.event_number = None
+    self.unknown_int = None
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:39]: I opt not to store unknowns

File: plaso/parsers/wincc.py [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+      raise errors.WrongParser(error_string)
+
+    # The first lines seem to  always follow the following format:
+    #
+    # 2019-05-27 10:05:43,405 INFO     Log starting ...
+    # 2019-05-27 10:05:43,405 INFO     | LogFileName   : C:\.....
+    # 2019-05-27 10:05:43,419 INFO     | LogFileCount  : 3
+
+    if line_number == 0:
+      expected_str = 'Log starting ...'+self.END_OF_LINE
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:82]: use join for string concatenation

File: tests/parsers/wincc.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        'unknown_int': 4,
+        'unknown_str': '',
+        'hostname': 'BMS001',
+        'source': 'CCWriteArchiveServer',
+        'message': ('[(null) 224]failed to insert into MSARCLONG with '
+                    '0x80004005(#0 \'2019-05-27 08:10:03.602\') MSG_STATE_COME')
+    }
+    event_data = storage_writer.GetAttributeContainerByIndex('event_data', 2)
+    self.CheckEventData(event_data, expected_event_values)
+
```
Comment (joachimmetz) [inline tests/parsers/wincc.py:75]: add 1 empty line

File: plaso/parsers/wincc.py [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+      if not values[1].endswith(expected_str):
+        error_string = 'Expected first line to end with "{0:s}"."{1!s}"'.format(
+            expected_str, values[1])
+        raise errors.WrongParser(error_string)
+
+    if line_number == 1:
+      expected_str = '| LogFileName'
+      if values[1].find(expected_str) < 0:
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:90]: instead of find could you use: `'| LogFileName' not in values[1]` ?

File: plaso/parsers/wincc.py [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+        error_string = 'Expected second line to contain "{0:s}"'.format(
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:91]: since expected_str is static why not hard code the error string?

File: plaso/parsers/wincc.py [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
+          and other components, such as storage and dfVFS.
+      line_number (int): number of the line the values were extracted from.
+      values (list[str]): values extracted from the line.
+      first_line (bool): True if this is first line from which values were
+          extracted.
+
+    Raises:
+      WrongParser: when the values cannot be parsed.
+    """
+
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:177]: remove empty line

File: plaso/parsers/wincc.py [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+    nb_of_values = len(values)
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:178]: don't use (non-standard) abbreviations 

nb => number

File: plaso/parsers/wincc.py [hunk 13]
```diff
@@ -0,0 +1,341 @@
...
+      # produces more values.
+      if first_line:
+        raise errors.WrongParser(error_string)
+
+      parser_mediator.ProduceExtractionWarning(error_string)
+
+    event_data = WinCCSysLogEventData()
+
+    try:
+      log_id = int(values[0])
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:193]: id => identifier

File: plaso/parsers/wincc.py [hunk 14]
```diff
@@ -0,0 +1,341 @@
...
+    except ValueError as exc:
+      error_string = (
+          'Type of first value ({0!s}) should be an int in line: {1:d}').format(
+              values[0], line_number)
+      self._ParseValuesFail(parser_mediator, first_line, exc, error_string)
+
+    try:
+      date_string = values[1]
+      time_string = values[2]
+      day_of_month, month, year = [int(elem) for elem in date_string.split('.')]
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:204]: elem => element

File: plaso/parsers/wincc.py [hunk 15]
```diff
@@ -0,0 +1,341 @@
...
+    except (TypeError, ValueError) as exception:
+      error_string = (
+          'Unable to parse time elements with error: '
+          '{0!s} on line {1:d} {2!s}').format(exception, line_number, values)
+      self._ParseValuesFail(
+          parser_mediator, first_line, exception, error_string)
+
+    try:
+      event_data.event_number = int(values[3])
+    except ValueError as exc:
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:226]: exc => exception

File: plaso/parsers/wincc.py [hunk 16]
```diff
@@ -0,0 +1,341 @@
...
+        first_line = False
+        line = line_reader.readline().strip()
+      except UnicodeDecodeError as exception:
+        parser_mediator.ProduceExtractionWarning(
+            'unable to read line: {0:d} with error: {1!s}'.format(
+                line_number, exception))
+        break
+
+
+manager.ParsersManager.RegisterParser(WinCCSysLogParser)
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:340]: instead of the individual calls to RegisterParser use RegisterParsers([])

**Review by joachimmetz — CHANGES_REQUESTED** [review]
Some initial nits and comments

File: plaso/parsers/wincc.py (hunk 1 above)
Comment (rgayon) [reply plaso/parsers/wincc.py:15]: done

File: plaso/parsers/wincc.py (hunk 2 above)
Comment (rgayon) [reply plaso/parsers/wincc.py:28]: done

File: plaso/parsers/wincc.py (hunk 3 above)
Comment (rgayon) [reply plaso/parsers/wincc.py:27]: It is called [WinCC](https://en.wikipedia.org/wiki/WinCC)

File: plaso/parsers/wincc.py (hunk 12 above)
Comment (rgayon) [reply plaso/parsers/wincc.py:178]: done

File: plaso/parsers/wincc.py (hunk 14 above)
Comment (rgayon) [reply plaso/parsers/wincc.py:204]: done

File: plaso/parsers/wincc.py (hunk 15 above)
Comment (rgayon) [reply plaso/parsers/wincc.py:226]: done

File: plaso/parsers/wincc.py (hunk 16 above)
Comment (rgayon) [reply plaso/parsers/wincc.py:340]: done

File: plaso/parsers/wincc.py (hunk 5 above)
Comment (rgayon) [reply plaso/parsers/wincc.py]: done

File: plaso/parsers/wincc.py [hunk 17]
```diff
@@ -0,0 +1,362 @@
...
+from plaso.parsers import manager
+
+
+class SIMATICS7EventData(events.EventData):
+  """SIMATIC S7 event data.
+
+  Attributes:
+    creation_time (dfdatetime.DateTimeValues): date and time the log entry
+        was created.
+    message (str): the message of the event.
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:20]: message has a special purpose in Plaso (see: https://plaso.readthedocs.io/en/latest/sources/user/Output-and-formatting.html#dynamic-output-module-fields) use body (or equiv instead)

File: plaso/parsers/wincc.py [hunk 18]
```diff
@@ -0,0 +1,362 @@
...
+
+  Attributes:
+    log_identifier (int): identifier for this log file.
+    creation_time (dfdatetime.DateTimeValues): date and time the log entry
+        was created.
+    event_number (int): a number specifying the type of event.
+    unknown_int (int): an integer with unknown meaning.
+    unknown_str (str): a string with unknown meaning (mostly empty).
+    hostname (str): the hostname of the machine logging the event.
+    source (str): which device generated the event.
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:43]: message, source have a special purpose in Plaso (see: https://plaso.readthedocs.io/en/latest/sources/user/Output-and-formatting.html#dynamic-output-module-fields) use body (or equiv instead)

hostname can be used but will be can be set to a value determined by pre-processing https://github.com/log2timeline/plaso/blob/main/plaso/output/mediator.py#L127

remove unknown_int and unknown_str

File: plaso/parsers/wincc.py [hunk 19]
```diff
@@ -0,0 +1,362 @@
...
+    message (str): the content of the log's message.
+  """
+
+  DATA_TYPE = 'wincc:sys_log:entry'
+
+  def __init__(self):
+    """Initializes event data."""
+    super(WinCCSysLogEventData, self).__init__(data_type=self.DATA_TYPE)
+
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:52]: remove empty line

File: plaso/parsers/wincc.py [hunk 20]
```diff
@@ -0,0 +1,362 @@
...
+      raise errors.WrongParser(error_string)
+
+    # The first lines seem to  always follow the following format:
+    #
+    # 2019-05-27 10:05:43,405 INFO     Log starting ...
+    # 2019-05-27 10:05:43,405 INFO     | LogFileName   : C:\.....
+    # 2019-05-27 10:05:43,419 INFO     | LogFileCount  : 3
+
+    if line_number == 0:
+      expected_str = 'Log starting ...'+self.END_OF_LINE
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:98]: expected_str => expected_string

**Comment by joachimmetz** [comment]
@rgayon you'll also need to update the formatter and timeliner configuration files.

File: plaso/parsers/wincc.py (hunk 17 above)
Comment (rgayon) [reply plaso/parsers/wincc.py:20]: done

File: plaso/parsers/wincc.py (hunk 19 above)
Comment (rgayon) [reply plaso/parsers/wincc.py:52]: done

**Comment by rgayon** [comment]
> @rgayon you'll also need to update the formatter and timeliner configuration files, also see: https://plaso.readthedocs.io/en/latest/sources/developer/How-to-write-a-parser.html#writing-a-parser

done

**Comment by joachimmetz** [comment]
> I really have no idea why my code keeps being mangled

what are you running into?

> done

thx, I'll have another look when time permits

File: data/formatters/windows.yaml [hunk 21]
```diff
@@ -81,6 +81,27 @@ short_source: 'REG'
 source: 'Task Cache Registry Key'
 ---
 type: 'conditional'
+data_type: 'wincc:simatic_s7:entry'
+message:
+- '{body}'
+short_message:
+- '{body}'
+short_source: 'WINCC'
```
Comment (joachimmetz) [inline data/formatters/windows.yaml]: Informational: I think the original intent here is that this should be LOG

File: plaso/parsers/wincc.py [hunk 22]
```diff
@@ -0,0 +1,345 @@
...
+            '{1:d}').format(hostname, line_number)
+        self._ParseValuesFail(
+            parser_mediator, first_line, exception, error_string)
+      if len(hostname) > 16:
+        error_string = (
+            'Hostname ({0!s}) can\'t be longer than 15 characters on line:'
+            '{1:d}').format(hostname, line_number)
+        self._ParseValuesFail(
+            parser_mediator, first_line, exception, error_string)
+      disallowed_characters = ['\\', '/', ':', '*', '?', '"', '<', '>', '|']
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:266]: make this a static class definition

File: plaso/parsers/wincc.py [hunk 23]
```diff
@@ -0,0 +1,345 @@
...
+    # 2019-05-27 10:05:43,419 INFO     | LogFileCount  : 3
+    if line_number == 0:
+      expected_string = 'Log starting ...'+self.END_OF_LINE
+      if not values[1].endswith(expected_string):
+        error_string = 'Expected first line to end with "{0:s}"."{1!s}"'.format(
+            expected_string, values[1])
+        raise errors.WrongParser(error_string)
+
+    if line_number == 1:
+      expected_string = '| LogFileName'
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:99]: make this a static class definition

File: plaso/parsers/wincc.py [hunk 24]
```diff
@@ -0,0 +1,345 @@
...
+      if values[1].find(expected_string) < 0:
+        error_string = 'Expected second line to contain "{0:s}"'.format(
+            expected_string)
+        raise errors.WrongParser(error_string)
+
+    if line_number == 2:
+      expected_string = '| LogFileCount'
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:106]: make this a static class definition

File: plaso/parsers/wincc.py [hunk 25]
```diff
@@ -0,0 +1,345 @@
...
+          'Unable to parse date time value with error: '
+          '{0!s} on line {1:d} {2!s}').format(exception, line_number, values)
+      parser_mediator.ProduceExtractionWarning(error_string)
+
+    event_data.body = ','.join(values[1:]).strip()
+
+    parser_mediator.ProduceEventData(event_data)
+
+  def ParseFileObject(self, parser_mediator, file_object):
+    """Parses a bodyfile file-like object.
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:128]: bodyfile ? copy/paste left over?

File: plaso/parsers/wincc.py [hunk 26]
```diff
@@ -0,0 +1,345 @@
...
+          'Type of first value ({0!s}) should be an int in line: {1:d}').format(
+              values[0], line_number)
+      self._ParseValuesFail(
+          parser_mediator, first_line, exception, error_string)
+
+    try:
+      date_string = values[1]
+      time_string = values[2]
+      day_of_month, month, year = [
+          int(element) for element in date_string.split('.')]
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:216]: add the integer base e.g. `int(element, 10)`

File: plaso/parsers/wincc.py [hunk 27]
```diff
@@ -0,0 +1,345 @@
...
+      )
+      # Unsure about whether the format is dependant on the system's Locale
+      date_time.is_local_time = True
+
+      event_data.creation_time = date_time
+
+    except (TypeError, ValueError) as exception:
+      error_string = (
+          'Unable to parse time elements with error: '
+          '{0!s} on line {1:d} {2!s}').format(exception, line_number, values)
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:232]: values => values[1:2] ?

File: plaso/parsers/wincc.py [hunk 28]
```diff
@@ -0,0 +1,345 @@
...
+      self._ParseValuesFail(
+          parser_mediator, first_line, exception, error_string)
+
+    try:
+      event_data.event_number = int(values[3])
+    except ValueError as exception:
+      error_string = (
+          'Type of event_number value ({0!s}) should be an int in line:'
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:240]: `an int` => `an integer` or `a decimal` ?

File: plaso/parsers/wincc.py [hunk 29]
```diff
@@ -0,0 +1,345 @@
...
+
+    except (TypeError, ValueError) as exception:
+      error_string = (
+          'Unable to parse time elements with error: '
+          '{0!s} on line {1:d} {2!s}').format(exception, line_number, values)
+      self._ParseValuesFail(
+          parser_mediator, first_line, exception, error_string)
+
+    try:
+      event_data.event_number = int(values[3])
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:237]: add integer base

File: plaso/parsers/wincc.py [hunk 30]
```diff
@@ -0,0 +1,345 @@
...
+
+  def _ParseValuesFail(
+      self, parser_mediator, first_line, parent_exception, error_string):
+    if first_line:
+      raise errors.WrongParser(error_string) from parent_exception
+
+    parser_mediator.ProduceExtractionWarning(error_string)
+
+  def ParseFileObject(self, parser_mediator, file_object):
+    """Parses a bodyfile file-like object.
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:296]: bodyfile ?

File: plaso/parsers/wincc.py [hunk 31]
```diff
@@ -0,0 +1,345 @@
...
+
+    try:
+      line = line_reader.readline().strip()
+    except UnicodeDecodeError as exception:
+      raise errors.WrongParser(
+          'unable to read line: {0:d} with error: {1!s}'.format(
+              line_number, exception))
+
+    while line:
+      if line.startswith('======>'):
```
Comment (joachimmetz) [inline plaso/parsers/wincc.py:325]: make this a static class definition

**Review by joachimmetz — CHANGES_REQUESTED** [review]
left some comments

File: plaso/parsers/wincc.py (hunk 22 above)
Comment (rgayon) [reply plaso/parsers/wincc.py:266]: done

File: plaso/parsers/wincc.py (hunk 23 above)
Comment (rgayon) [reply plaso/parsers/wincc.py:99]: done

File: plaso/parsers/wincc.py (hunk 26 above)
Comment (rgayon) [reply plaso/parsers/wincc.py:216]: done (4 times)

File: plaso/parsers/wincc.py (hunk 27 above)
Comment (rgayon) [reply plaso/parsers/wincc.py:232]: done

File: plaso/parsers/wincc.py (hunk 28 above)
Comment (rgayon) [reply plaso/parsers/wincc.py:240]: done

File: plaso/parsers/wincc.py (hunk 31 above)
Comment (rgayon) [reply plaso/parsers/wincc.py:325]: done

**Comment by rgayon** [comment]
Looks like there is an issue with building docs

**Comment by joachimmetz** [comment]
rebased with upstream to not have to review 194 files instead of 7

**Review by joachimmetz — APPROVED** [review]
LGTM

**Pull request merged**